core/host_api/impl/crypto_extension.cpp

```diff
@@ -562,4 +569,130 @@ namespace kagome::host_api {
     auto buffer = scale::encode(ResultType(public_key.value())).value();
     return getMemory().storeBuffer(buffer);
   }
+
+  runtime::WasmSpan CryptoExtension::ext_crypto_ecdsa_public_keys_version_1(
+      runtime::WasmSize key_type) {
+    using ResultType = std::vector<crypto::EcdsaPublicKey>;
+    static const auto error_result(scale::encode(ResultType{}).value());
+
+    auto key_type_id =
+        static_cast<crypto::KeyTypeId>(getMemory().load32u(key_type));
+    if (!crypto::isSupportedKeyType(key_type_id)) {
+      logger_->warn("key type '{}' is not officially supported",
+                    common::int_to_hex(key_type_id, 8));
+    }
+    auto public_keys = crypto_store_->getEcdsaPublicKeys(key_type_id);
+    if (not public_keys) {
+      auto msg = fmt::format("error loading public keys: {}",
+                             public_keys.error().message());
+      throw std::runtime_error(msg);
+    }
+
+    auto buffer = scale::encode(public_keys.value()).value();
+    SL_TRACE_FUNC_CALL(logger_, public_keys.value().size(), key_type_id);
+
+    return getMemory().storeBuffer(buffer);
+  }
+
+  runtime::WasmSpan CryptoExtension::ext_crypto_ecdsa_sign_version_1(
+      runtime::WasmSize key_type,
+      runtime::WasmPointer key,
+      runtime::WasmSpan msg) {
+    using ResultType = std::optional<crypto::EcdsaSignature>;
+
+    auto key_type_id =
+        static_cast<crypto::KeyTypeId>(getMemory().load32u(key_type));
+    if (!crypto::isSupportedKeyType(key_type_id)) {
+      logger_->warn("key type '{}' is not officially supported",
+                    common::int_to_hex(key_type_id, 8));
+    }
+    auto public_buffer = getMemory().loadN(key, sizeof(crypto::EcdsaPublicKey));
+    auto [msg_data, msg_len] = runtime::PtrSize(msg);
+    auto msg_buffer = getMemory().loadN(msg_data, msg_len);
+
+    crypto::EcdsaPublicKey pk;
+    std::copy(public_buffer.begin(), public_buffer.end(), pk.begin());
+    auto key_pair = crypto_store_->findEcdsaKeypair(key_type_id, pk);
+    if (!key_pair) {
+      logger_->error("failed to find required key");
+      auto error_result = scale::encode(ResultType(std::nullopt)).value();
+      return getMemory().storeBuffer(error_result);
+    }
+
+    auto sign = ecdsa_provider_->sign(msg_buffer, key_pair.value().secret_key);
+    if (!sign) {
+      logger_->error("failed to sign message, error = {}",
+                     sign.error().message());
+      std::terminate();
```

Comment:
Ouch! **terminate**? If it is reasonable, change log to critical

Reply:
Probably, throwing exception is more appropriate here.

Reply:
Again. like ed25519 and sr25519. Is it wrong? Should it be there as well?

Reply:
Yep! It good idea to make code better.
It was implemented with other logger early, and probably errors had flushing immediately, not sure. But using crilital level very actual now.